### PR TITLE
[PDS-158510] Improved Detection of Orphaned Sentinels

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- PDS-158510: some users were running into issues with orphaned sentinel skip not working.

**Implementation Description (bullets)**:
- In #3610 detection and skipping of sweep sentinels was added. However, there is an edge case: if a transaction T1 reads an orphaned sentinel S, for which an _uncommitted_ T2 has written to the KVS, then T1 doesn't recognise S as orphaned even though the value T2 covered it with is uncommitted.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Added a new group of tests.

**Concerns (what feedback would you like?)**:
Not much. Is perf a concern?

**Where should we start reviewing?**: Probably start with the tests

**Priority (whenever / two weeks / yesterday)**: yesterday

### CODE REVIEW

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 1]
```diff
@@ -1438,6 +1443,190 @@ public void checkImmutableTsLockAfterReadsForConservativeIfFlagIsSet() {
...
+                .isInstanceOf(TransactionFailedRetriableException.class)
+                .hasMessageContaining("Tried to read a value that has been deleted.");
+
+        Transaction retryTxn = txManager.createNewTransaction();
+        assertThatCode(() -> retryTxn.get(TABLE_SWEPT_CONSERVATIVE, ImmutableSet.of(TEST_CELL)))
+                .doesNotThrowAnyException();
+    }
+
+    @Test
+    public void getSweepSentinelUnderEarlyUncommittedWriteDoesNotThrow() {
```
Comment (gmaretic) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1473]: should swap the names of this and the following test

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 2]
```diff
@@ -1438,6 +1443,190 @@ public void checkImmutableTsLockAfterReadsForConservativeIfFlagIsSet() {
...
+        t2.commit();
+
+        assertThatThrownBy(() -> t1.get(TABLE_SWEPT_CONSERVATIVE, ImmutableSet.of(TEST_CELL, TEST_CELL_2)))
+                .isInstanceOf(TransactionFailedRetriableException.class)
+                .hasMessageContaining("Tried to read a value that has been deleted.");
+    }
+
+    @SuppressWarnings("unchecked") // ArgumentCaptor
+    @Test
+    public void getSentinelValuesStressTest() {
```
Comment (gmaretic) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1565]: nice 👍

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 3]
```diff
@@ -1438,6 +1443,190 @@ public void checkImmutableTsLockAfterReadsForConservativeIfFlagIsSet() {
...
+
+        List<Iterable<Long>> stressTestRequests = captor.getAllValues();
+        assertThat(stressTestRequests).hasSize(100);
+        for (int index = 1; index < stressTestRequests.size(); index++) {
+            List<Long> previousTimestamps = StreamSupport.stream(
+                            stressTestRequests.get(index - 1).spliterator(), false)
+                    .collect(Collectors.toList());
+            assertThat(stressTestRequests.get(index)).hasSizeLessThan(previousTimestamps.size());
+        }
+    }
```
Comment (gmaretic) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1616]: assert that retry succeeds?

**Review by gmaretic — APPROVED** [review]
Paired on most of this.

A few nits. I think perf is fine, took a look at the metrics for encountering naked sentinels and it's not very common. The reads are going to dominate the perf of this either way and there is no way around that.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1473]: ah, you're right!

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java (hunk 3 above)
Comment (jeremyk-91) [reply atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:1616]: Yep, makes sense - added.

**Pull request merged**